Issue #13 🎫: Game class

ITurres opened the pull request:
# 🚩I have made the following changes to complete Issue #13

## Here is a summary of what has been done

### Type of **`files`** changed/added

![My Skills](https://skillicons.dev/icons?i=ruby)

### In this new Branch, I have **➕ADDED➕** the succeeding **\*📂 folders** and/or **\*\*📄 files** to complete this ticket's [requirements](https://github.com/ITurres/catalog_of_things/issues/13)

# \*📂 root/

- ## \*📂 lib/

  - **\*\*📄 game.rb** ➕

    - I have added all the attributes and methods required.

      - @demesameneshoa  **❗Please consider reading the comments on the `game.rb` file** it would be great to have a second opinion 🙏🏼.

### Thank you for taking the time to review this PR ⭐

---

If you need additional information or have any questions, please don't hesitate to contact me on Slack as _**Arturo (Arthur) Emanuel Guerra Iturres**_. I'll be happy to help you out. 🎯

---

ITurres committed:
Add 'game.rb / Game' class required attr and methods

demesameneshoa commented on lib/game.rb:
```diff
@@ -0,0 +1,28 @@
+    return false if @last_played_at.nil? || @last_played_at !~ /\d{4}-\d{2}-\d{2}/
+
+    two_years_in_days = 365 * 2 # ! 2 years in days - not counting leap years and/or 365.25 days per year (average)
+    last_played_at_date_year = Date.parse(@last_played_at).year
+
+    (Date.today - last_played_at_date_year) > two_years_in_days
```
Here the last_played_at_date_year is going to to get the year value where as when comparing it with date.today we are going to be comparing date and year. kindly check this as it might be an issue

ITurres replied:
Hi @demesameneshoa! Thank you, partner! absolutely, It should be comparing it with the `.year` method. I will make the changes right away 🎯.

demesameneshoa commented on lib/game.rb:
```diff
@@ -0,0 +1,28 @@
+  attr_reader :multiplayer, :last_played_at
+
+  def initialize(last_played_at, multiplayer)
+    super(last_played_at)
+    @last_played_at = last_played_at
```
There seems to be a redundancy in initialization one in super and one in the sub class, wouldn't it be enough to use the super initialization?

ITurres replied:
In this case, we need the information on the `last_played_at` argument to be sent to the parent class, since this one will serve as the parent `publish_date` param.

 ~~However, I have the feeling that we need to take `publish_date` as an argument in the `Game` class, @demesameneshoa  What do you think❓ given the fact that both are different names and might be used for different things?~~

Update: Now the constructor will take `publish_date` and pass it to super as well as taking the other associated class instances.

In the same way, we need `last_played_at` as an instance variable to be used in the methods 🤠.

demesameneshoa replied:
@ITurres you have raised a valid request, I am sorry i haven't taken your point into consideration. This is Approved

demesameneshoa reviewed (changes requested):
Hi @ITurres , 
Changes Required  🟡

You have done such an amazing job! but i have a few comments💬  


<img src="https://media.giphy.com/media/ReyD4OdRiHeLDCRaEP/giphy.gif"/>

## Highlights
✅ The `game` class is created
✅ The Class is initialized properly
✅ The genre class takes instance of `Item` class

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._



Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table  width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Demes Ameneshoa</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/demesameneshoa">
        <img src="https://img.shields.io/badge/github-%23121011.svg?style=for-the-badge&logo=github&logoColor=white" alt="GitHub">
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/demesameneshoa/">
        <img src="https://img.shields.io/badge/linkedin-%230077B5.svg?style=for-the-badge&logo=linkedin&logoColor=white" alt="LinkedIn">
      </a>
    </td>
    <td align="center">
      <a href="https://microverse-students.slack.com/team/U051695H9RS">
        <img src="https://img.shields.io/badge/Slack-4A154B?style=for-the-badge&logo=slack&logoColor=white" alt="Slack">
      </a>
    </td>
    <td align="center">
      <a href="mailto:demesameneshoa@gmail.com?subject=Microverse%20Code%20Review">
        <img src="https://img.shields.io/badge/Gmail-D14836?style=for-the-badge&logo=gmail&logoColor=white" alt="Email">
      </a>
    </td>
  </tr>
</table>

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

ITurres committed:
Fix: 'game.rb' Access 'year' method for 'current_year' var

ITurres committed:
Feat: Enhance 'Game' class constructor to:

- Accept:
    - A item object with the instantiation of: Genre, Author and Label.
    - super publish_date with default value of nil
    - last_played_at with default value of nil

demesameneshoa reviewed (approved):
Hi @ITurres , 

Your ticket 🎫  is complete!  🟢 There is nothing else to say other than it's time to merge it :shipit: 
Congratulations! 🎉 

<img src="https://media.giphy.com/media/a0h7sAqON67nO/giphy.gif"/>

## Highlights
✅ The requested changes are implemented
✅ No linters errors
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._



Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table  width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Demes Ameneshoa</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/demesameneshoa">
        <img src="https://img.shields.io/badge/github-%23121011.svg?style=for-the-badge&logo=github&logoColor=white" alt="GitHub">
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/demesameneshoa/">
        <img src="https://img.shields.io/badge/linkedin-%230077B5.svg?style=for-the-badge&logo=linkedin&logoColor=white" alt="LinkedIn">
      </a>
    </td>
    <td align="center">
      <a href="https://microverse-students.slack.com/team/U051695H9RS">
        <img src="https://img.shields.io/badge/Slack-4A154B?style=for-the-badge&logo=slack&logoColor=white" alt="Slack">
      </a>
    </td>
    <td align="center">
      <a href="mailto:demesameneshoa@gmail.com?subject=Microverse%20Code%20Review">
        <img src="https://img.shields.io/badge/Gmail-D14836?style=for-the-badge&logo=gmail&logoColor=white" alt="Email">
      </a>
    </td>
  </tr>
</table>

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

ITurres commented:
@demesameneshoa Thank you Demes! 🥳
